wgengine/netstack/netstack.go

```diff
@@ -112,86 +79,180 @@ func Impl(logf logger.Logf, tundev *tstun.TUN, e wgengine.Engine, mc *magicsock.
-			full = append(full, hdrTransport.View()...)
-			full = append(full, pkt.Data.ToView()...)
-
-			logf("XXX packet Write out: % x", full)
-			if err := tundev.InjectOutbound(full); err != nil {
-				log.Printf("netstack inject outbound: %v", err)
-				return
-			}
+func (ns *Impl) dialTCP(address string) (*gonet.TCPConn, error) {
+	remoteIPStr, remotePortStr, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, errors.New("could not parse IP:port: " + err.Error())
+	}
+	remotePort, _ := strconv.Atoi(remotePortStr)
+	remoteIP := net.ParseIP(remoteIPStr)
+	remoteAddress := tcpip.FullAddress{
+		NIC:  nicID,
+		Addr: tcpip.Address(remoteIP),
+		Port: uint16(remotePort),
+	}
+	var ipType tcpip.NetworkProtocolNumber
+	if remoteIP.To4() != nil {
+		ipType = ipv4.ProtocolNumber
+	} else {
+		ipType = ipv6.ProtocolNumber
+	}
+
+	return gonet.DialTCP(ns.ipstack, remoteAddress, ipType)
+}
 
+func (ns *Impl) injectOutbound() {
+	for {
+		packetInfo, ok := ns.linkEP.ReadContext(context.Background())
+		if !ok {
+			ns.logf("XXX ReadContext-for-write = ok=false")
+			continue
 		}
-	}()
-
-	tundev.PostFilterIn = func(p *packet.Parsed, t *tstun.TUN) filter.Response {
-		var pn tcpip.NetworkProtocolNumber
-		switch p.IPVersion {
-		case 4:
-			pn = header.IPv4ProtocolNumber
-		case 6:
-			pn = header.IPv6ProtocolNumber
+		pkt := packetInfo.Pkt
+		hdrNetwork := pkt.NetworkHeader()
+		hdrTransport := pkt.TransportHeader()
+
+		full := make([]byte, 0, pkt.Size())
+		full = append(full, hdrNetwork.View()...)
+		full = append(full, hdrTransport.View()...)
+		full = append(full, pkt.Data.ToView()...)
+
+		ns.logf("XXX packet Write out: % x", full)
+		if err := ns.tundev.InjectOutbound(full); err != nil {
+			log.Printf("netstack inject outbound: %v", err)
+			return
 		}
-		logf("XXX packet in (from %v): % x", p.Src, p.Buffer())
-		vv := buffer.View(append([]byte(nil), p.Buffer()...)).ToVectorisedView()
-		packetBuf := stack.NewPacketBuffer(stack.PacketBufferOptions{
-			Data: vv,
-		})
-		linkEP.InjectInbound(pn, packetBuf)
-		return filter.Accept
+
 	}
-	return nil
 }
 
-func echo(c *gonet.TCPConn, e wgengine.Engine, mc *magicsock.Conn) {
-	defer c.Close()
-	src, _ := netaddr.FromStdIP(c.RemoteAddr().(*net.TCPAddr).IP)
-	who := ""
-	if n, u, ok := mc.WhoIs(src); ok {
-		who = fmt.Sprintf("%v from %v", u.DisplayName, n.Name)
+func (ns *Impl) injectInbound(p *packet.Parsed, t *tstun.TUN) filter.Response {
+	var pn tcpip.NetworkProtocolNumber
+	switch p.IPVersion {
+	case 4:
+		pn = header.IPv4ProtocolNumber
+	case 6:
+		pn = header.IPv6ProtocolNumber
+	}
+	ns.logf("XXX packet in (from %v): % x", p.Src, p.Buffer())
+	vv := buffer.View(append([]byte(nil), p.Buffer()...)).ToVectorisedView()
+	packetBuf := stack.NewPacketBuffer(stack.PacketBufferOptions{
+		Data: vv,
+	})
+	ns.linkEP.InjectInbound(pn, packetBuf)
+	return filter.Accept
+}
+
+func (ns *Impl) acceptTCP(r *tcp.ForwarderRequest) {
+	ns.logf("XXX ForwarderRequest: %v", r)
```

Comment:
maybe replace all these XXX with `[v2]` to annotate their log level?  Then `tailscaled --verbose=2` would show them